### ISSUE TITLE
SNI routing by egressgateway doesn't work with ISTIO_MUTUAL

### ISSUE THREAD

**Issue by ronenschafferibm** [istio/istio#7415]
**Describe the bug**
SNI routing by egressgateway doesn't work with ISTIO_MUTUAL.

**Expected behavior**
Directing HTTPS traffic through the egressgateway when tls mode is set to ISTIO_MUTUAL should work.

**Steps to reproduce the bug**
Following issue #7361, I tried to configure scenario 3a with ISTIO_MUTUAL.
The scenario is enabling mTLS on the traffic from an application pod and the egress gateway.

First, I tried replacing the tls mode of the Gateway and the DestinationRule from MUTUAL to ISTIO_MUTUAL.

```
cat <<EOF | istioctl create -f -
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  ports:
  - number: 443
    name: tls
    protocol: TLS
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: google
spec:
  hosts:
  - www.google.com
  ports:
  - number: 443
    name: tls
    protocol: TLS
  resolution: DNS
---
  apiVersion: networking.istio.io/v1alpha3
  kind: Gateway
  metadata:
    name: istio-egressgateway
  spec:
    selector:
      istio: egressgateway
    servers:
    - port:
        number: 443
        name: tls-cnn
        protocol: TLS
      hosts:
      - edition.cnn.com
      tls:
        mode: ISTIO_MUTUAL
    - port:
        number: 443
        name: tls-google
        protocol: TLS
      hosts:
      - www.google.com
      tls:
        mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: cnn-to-egressgateway
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - mesh
  tls:
  - match:
    - gateways:
      - mesh
      port: 443
      sni_hosts:
      - edition.cnn.com
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: cnn
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: google-to-egressgateway
spec:
  hosts:
  - www.google.com
  gateways:
  - mesh
  tls:
  - match:
    - gateways:
      - mesh
      port: 443
      sni_hosts:
      - www.google.com
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: google
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - istio-egressgateway
  tcp:
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: google
spec:
  hosts:
  - www.google.com
  gateways:
  -  istio-egressgateway
  tcp:
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
    route:
    - destination:
        host: www.google.com
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
   name: set-sni-mtls-for-egressgateway
spec:
  host: istio-egressgateway.istio-system.svc.cluster.local
  subsets:
      - name: cnn
        trafficPolicy:
          loadBalancer:
            simple: ROUND_ROBIN
          portLevelSettings:
          - port:
              number: 443
            tls:
              mode: ISTIO_MUTUAL
              sni: edition.cnn.com
      - name: google
        trafficPolicy:
          loadBalancer:
            simple: ROUND_ROBIN
          portLevelSettings:
          - port:
              number: 443
            tls:
              mode: ISTIO_MUTUAL
              sni: www.google.com
EOF
```

which returns the following error:
```
unknown value "ISTIO_MUTUAL" for enum istio.networking.v1alpha3.Server_TLSOptions_TLSmode
```
The reason seems to be the difference between [`TLSSettings.TLSmode`](https://github.com/istio/api/blob/4257ff7bbb6ea82e5a525d33bbc8d81492d9dba0/networking/v1alpha3/destination_rule.proto#L494) and [`TLSOptions.TLSmode`](https://github.com/istio/api/blob/4257ff7bbb6ea82e5a525d33bbc8d81492d9dba0/networking/v1alpha3/gateway.proto#L252)

The former is used by `DestinationRule` and includes `ISTIO_MUTUAL` while the latter is used by `Gateway` and lacks it.

Second, I tried replacing the tls mode only in the DestinationRule.

```
cat <<EOF | istioctl create -f -
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  ports:
  - number: 443
    name: tls
    protocol: TLS
  resolution: DNS
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: google
spec:
  hosts:
  - www.google.com
  ports:
  - number: 443
    name: tls
    protocol: TLS
  resolution: DNS
---
  apiVersion: networking.istio.io/v1alpha3
  kind: Gateway
  metadata:
    name: istio-egressgateway
  spec:
    selector:
      istio: egressgateway
    servers:
    - port:
        number: 443
        name: tls-cnn
        protocol: TLS
      hosts:
      - edition.cnn.com
      tls:
        mode: MUTUAL
        serverCertificate: /etc/certs/cert-chain.pem
        privateKey: /etc/certs/key.pem
        caCertificates: /etc/certs/root-cert.pem
    - port:
        number: 443
        name: tls-google
        protocol: TLS
      hosts:
      - www.google.com
      tls:
        mode: MUTUAL
        serverCertificate: /etc/certs/cert-chain.pem
        privateKey: /etc/certs/key.pem
        caCertificates: /etc/certs/root-cert.pem
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: cnn-to-egressgateway
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - mesh
  tls:
  - match:
    - gateways:
      - mesh
      port: 443
      sni_hosts:
      - edition.cnn.com
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: cnn
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: google-to-egressgateway
spec:
  hosts:
  - www.google.com
  gateways:
  - mesh
  tls:
  - match:
    - gateways:
      - mesh
      port: 443
      sni_hosts:
      - www.google.com
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: google
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: cnn
spec:
  hosts:
  - edition.cnn.com
  gateways:
  - istio-egressgateway
  tcp:
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
    route:
    - destination:
        host: edition.cnn.com
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: google
spec:
  hosts:
  - www.google.com
  gateways:
  -  istio-egressgateway
  tcp:
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
    route:
    - destination:
        host: www.google.com
        port:
          number: 443
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
   name: set-sni-mtls-for-egressgateway
spec:
  host: istio-egressgateway.istio-system.svc.cluster.local
  subsets:
      - name: cnn
        trafficPolicy:
          loadBalancer:
            simple: ROUND_ROBIN
          portLevelSettings:
          - port:
              number: 443
            tls:
              mode: ISTIO_MUTUAL
              sni: edition.cnn.com
      - name: google
        trafficPolicy:
          loadBalancer:
            simple: ROUND_ROBIN
          portLevelSettings:
          - port:
              number: 443
            tls:
              mode: ISTIO_MUTUAL
              sni: www.google.com
EOF
```

test:
```
kubectl exec -it $(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name}) -c sleep -- curl -sL -o /dev/null -w "%{http_code}\n" https://edition.cnn.com; kubectl exec -it $(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name}) -c sleep -- curl -sL -o /dev/null -w "%{http_code}\n" https://www.google.com ; kubectl exec -it $(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name}) -c sleep -- curl -sL -o /dev/null -w "%{http_code}\n" https://www.wikipedia.org
```

actual output:
```
000
command terminated with exit code 35
000
command terminated with exit code 35
000
command terminated with exit code 51
```

expected output:
```
200
200
000
```

cleanup:
```
istioctl delete serviceentry cnn google
istioctl delete virtualservice cnn google cnn-to-egressgateway google-to-egressgateway
istioctl delete gateway istio-egressgateway
istioctl delete destinationrule set-sni-mtls-for-egressgateway
```

**Version**
https://github.com/istio/istio/tree/8a1fb344ea162472d42059077944f3aa0ec971fb

**Is Istio Auth enabled or not?**
yes

**Environment**
Kubernetes

**Comment by rshriram** [comment]
It won't work for gateways. You have to manually specify the certs for gateway. The fix I added was only for destination rules.

**Comment by ronenschafferibm** [comment]
Solved by #7423
